proxy/http/remap/RemapPluginInfo.cc

```diff
@@ -21,60 +21,249 @@
   limitations under the License.
  */
 
+#include <unistd.h>
+
 #include "RemapPluginInfo.h"
 #include "tscore/ink_string.h"
 #include "tscore/ink_memory.h"
+#include "tscore/ink_apidefs.h"
 
-RemapPluginInfo::List RemapPluginInfo::g_list;
+#include "RemapPluginInfo.h"
+#ifdef PLUGIN_DSO_TESTS
+#include "unit-tests/plugin_testing_common.h"
+#else
+#include "tscore/Diags.h"
+#endif
 
-RemapPluginInfo::RemapPluginInfo(ts::file::path &&library_path) : path(std::move(library_path)) {}
+/**
+ * @brief helper function that returns the function address from the plugin DSO
+ *
+ * There can be valid defined DSO symbols that are NULL
+ * but when it comes to functions we can assume that
+ * if not defined we can return nullptr and a valid address if the are defined.
+ * @param symbol function symbol name
+ * @param error error messages in case of symbol is not found
+ * @return function address or nullptr if not found.
+ */
+template <class T>
+T *
+RemapPluginInfo::getFunctionSymbol(const char *symbol)
+{
+  std::string error; /* ignore the error, return nullptr if symbol not defined */
+  void *address = nullptr;
+  getSymbol(symbol, address, error);
+  return reinterpret_cast<T *>(address);
+}
 
-RemapPluginInfo::~RemapPluginInfo()
+std::string
+RemapPluginInfo::missingRequiredSymbolError(const std::string &pluginName, const char *required, const char *requiring)
 {
-  if (dl_handle) {
-    dlclose(dl_handle);
+  std::string error;
+  error.assign("plugin ").append(pluginName).append(" missing required function ").append(required);
+  if (requiring) {
+    error.append(" if ").append(requiring).append(" is defined");
   }
+  return error;
```

Comment:
`return std::move(error);`

Reply:
I wonder if this is necessary / better. 

The intent here was to rely on "named return value optimization" (NRVO [copy elision](https://en.cppreference.com/w/cpp/language/copy_elision)). 

> In a return statement, when the operand is the name of a non-volatile object with automatic storage duration, which isn't a function parameter or a catch clause parameter, and which is of the same class type (ignoring cv-qualification) as the function return type. This variant of copy elision is known as NRVO, "named return value optimization".

My understanding is that while `std::move` may be efficient, "return value optimization" typically results in faster code because the object is constructed in place and nothing is moved or copied.